packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -129,9 +182,27 @@ export class ConnectionStateHandler extends EventEmitterWithErrorHandling<IConne
                 }
             }
             this._clientId = this.pendingClientId;
+            // Set _clientSentOps to false as this is a fresh connection.
+            this._clientSentOps = false;
         } else if (value === ConnectionState.Disconnected) {
             // Important as we process our own joinSession message through delta request
             this._pendingClientId = undefined;
+            // Only wait for "leave" message if we have some outstanding ops and the client was write client as
+            // server would not accept ops from read client. Also check if the timer is not already running as we
+            // could receive "Disconnected" event multiple times without getting connected and in that case we
+            // don't want to reset the timer as we still want to wait on original client which started this timer.
+            // We also check the dirty state of this connection as we only want to wait for the client leave of the
+            // client which created the ops. This helps with situation where a client disconnects immediately after
+            // getting connected without sending any ops(from previous client). In this case, we would join as write
+            // because there would be a diff between client seq number and clientSeqNumberObserved but then we don't
+            // want to wait for newly disconnected client to leave as it has not sent any ops yet.
+            if (this.handler.shouldClientJoinWrite()
+                && this.clientConnectionMode === "write"
+                && this.prevClientLeftTimer.hasTimer === false
+                && this._clientSentOps
```

Comment:
Isn't this._clientSentOps always false here? In a sense that once we lost connection we set it to false.
We will not send any more ops until we transition to connected state, and we transition to connected state only when we go through this IF check, which sounds like should never happen?

Given shouldClientJoinWrite() should incorporate the fact that it returns true only when there are non-acked ops (would be great to reflect it in the name), do we need this._clientSentOps at all?
This assumes that indeed shouldClientJoinWrite() would keep returning true until we eventually get to connected event (which I'm not sure is true, worth double checking)

Reply:
BTW, should we have this logic (start timer) only on connecting transition?
Or at least when timer fires, differentiate in logging if we are connected to web socket or not at that moment.
Otherwise if I lost connection and stay forever disconnected, timer would expire and it would look like leave op never materialized in 90 seconds, when in fact we just do not know as we have no way to check it. It would be correct, but screw our metrics and decision making on what timeout number to use.

Reply:
this._clientSentOps: we set it to false on Connected event as we want to see whether new client sent any ops or not.

Reply:
We reset the clientsenumber on reconnection when we submit ops and previous client id and client id of ops mismatches.
So consider the case on reconnection after waiting for leave op, new client did not submitted any ops and got disconnected and in that case shouldClientJoinWrite() would keep returning true but then we don't want to wait for this client leave op as it didn't submitted any ops. This is kind of a edge case.

Reply:
I think we reduce waiting in case stated above.

Reply:
Do you mean we already do? How? I do not see any place where we reset this._clientSentOps to false while maintaining active connection (i.e. as response to receiving all acks). That's what shouldClientJoinWrite() accounts correctly, but this code basically ignores that signal by adding this._clientSentOps check, no?

Reply:
We don't need to set this._clientSentOps to false while having active connection. We just need it for once when the first op is sent. After first op is sent, this becomes true until next Conencted connection. Here is the case where we need it: https://github.com/microsoft/FluidFramework/pull/5484#discussion_r597993411
So when the client joins and gets disconnected immediately before sending any left over ops from pervious connection, clientShouldJoinWrtie would return true but still we don't want to wait for leave op. In this case, this._clientSentOps would help.
This condition is in AND, so it cannot increase cases where we wait for timer on disconnected event.

Reply:
Ok, my logic was a bit off. It does not increase waiting. But I think it's no-op to have this._clientSentOps tracking. 
Basically this.handler.shouldClientJoinWrite() === true follows that this._clientSentOps  === true.
And thus having this._clientSentOps is not adding or changing anything.

Reply:
So this might not be true:
 this.handler.shouldClientJoinWrite() === true follows that this._clientSentOps === true.
If we look in deltaManager when we reset the clientSeqNumbers on submit of ops.
`public submit(message)
        if (this.lastSubmittedClientId !== this.connection?.clientId) {
            this.lastSubmittedClientId = this.connection?.clientId;
            this.clientSequenceNumber = 0;
            this.clientSequenceNumberObserved = 0;
        }
`
So when the client joins and gets disconnected immediately before sending any left over ops from pervious connection, clientShouldJoinWrtie would return true but still we don't want to wait for leave op.

Reply:
There should be no ops sent if shouldClientJoinWrite() returned true and we did not transition to connected state.
If that were to happen, then we are screwed. So once we disconnect, we will wait never hit submit() until the code you are changing tells (though converting to connected state) that ops could be sent.

I believe runtime and container have checks that ops can't be sent in non-connected state, they should short-circuit such attempts.

The reason why this invariant is important: deltamanager uses this invariant to decide what mode to connect to. It will use "write" as long as shouldClientJoinWrite() returns true. And your code will not transition to connected in such state until we see leave or timeout. If shouldClientJoinWrite() would reset through multiple web socket connections transitions (but did not see our own leave), then DM will connect as "read" and container would immediately transition to "connected" and container runtime would immediately sent ops. That would always result on duplicate ops. So maintaining this invariant is super important and we should be able to fully rely on it.

Reply:
Resolving as per the discussion.